pylhc/calibration/beta.py

```diff
@@ -0,0 +1,275 @@
+import numpy as np
+import pandas as pd
+from typing import Tuple, List, Dict
+
+from omc3.utils import logging_tools
+from omc3.optics_measurements.constants import (
+    AMP_BETA_NAME,
+    BETA,
+    BETA_NAME,
+    ERR,
+    EXT,
+    S,
+)
+from pylhc.constants.calibration import (
+    BPMS,
+    BETA_STAR_ESTIMATION,
+    IPS,
+    LABELS,
+    TFS_INDEX,
+)
+from pylhc.constants.general import PLANES
+import tfs
+
+
+LOG = logging_tools.get_logger(__name__)
+
+
+def _get_beta_fit(
+    bpms: List[str],
+    beta_phase_tfs: pd.DataFrame,
+    plane: str
+) -> Tuple[pd.Series, pd.Series]:
+    """
+    This function returns a fit of the given beta values along with the
+    associated error.
+
+    Args:
+        positions (pd.Series): Positions of the BPMs to be fitted.
+        beta_values (pd.Series): Values of the BPMs to be fitted.
+        beta_err (pd.Series): Associated errors to the values.
+
+    Returns:
+        Tuple[pd.Series, pd.Series]: The elements returned are the values of
+        the fit of the beta values and the associated error.
+    """
+
+    def beta_function(x, a, b):
+        return a + ((x - b) ** 2) / a
+
+    def err_function(x, popt, pcov):
+        sa, sb, sab = pcov[0,0], pcov[1,1], pcov[0,1]
+        a, b = popt[0], popt[1]
+
+        beta_err = ((a**2 - (x-b)**2) / a**2)**2 * sa
+        beta_err += 4 * ((x-b) / a)**2 * sb
+        beta_err -= 4 * (x-b) * (a**2 - (x-b)**2) / a**3 * sab
+        return beta_err
+    
+    positions = beta_phase_tfs.reindex(bpms)[f"{S}"]
+    beta_phase = beta_phase_tfs.reindex(bpms)[f"{BETA}{plane}"]
+    beta_phase_err = beta_phase_tfs.reindex(bpms)[f"{ERR}{BETA}{plane}"]
+
+    # Get the rough IP position and beta star for the initial values
+    ip_position = (positions[-1] - positions[0]) / 2
+    initial_values = (BETA_STAR_ESTIMATION, ip_position)
+
+    # Get the curve fit for the expected parabola
+    valid = ~(np.isnan(positions) | np.isnan(beta_phase))
+
+    additional_args = {}
+    if sum(beta_phase_err[valid]) != 0:
+        additional_args = {'sigma': beta_phase_err[valid]}
+
+    popt, pcov = curve_fit(
+        beta_function,
+        positions[valid],
+        beta_phase[valid],
+        p0=initial_values,
+        maxfev=1000000,
+        absolute_sigma=True,
+        **additional_args,
+    )
+    
+    # Get the error from the covariance matrix
+    perr = np.sqrt(np.diag(pcov))
+
+    # Get the fitted beta and add the errors to get min/max values
+    beta_fit = beta_function(positions[valid], *popt)
+    beta_fit_err = err_function(positions[valid], popt, pcov)
+
+    import matplotlib.pyplot as plt
+    fig, ax = plt.subplots()
+    bfit, = ax.plot(positions[valid], beta_fit[valid])
+    b_meas, = ax.plot(positions[valid], beta_phase[valid], marker='o')
+
+    ax.errorbar(positions[valid], beta_fit[valid], yerr=beta_fit_err)
+    ax.legend((bfit, b_meas), ('β fit', 'β^φ'), loc='upper right', shadow=True)
+
+    plt.show()
```

Comment:
Is this here on purpose or leftover from debugging

Reply:
oh! good thing you saw it